### ISSUE TITLE
Support for Firefox

### ISSUE THREAD

**Issue by adius** [pavel-agarkov/Midnight-Lizard#19]
With Firefox 57, aka Quantum, the time has come to switch back, but I'm really missing Midnight Lizard. 😢

Any chance you'll also provide support for Firefox in the future?

**Comment by pavel-agarkov** [comment]
@adius  Yes! Definitely will. It is the first thing in my plan for 2018.

**Comment by rtlewis1** [comment]
Hello, best extension developer on the planet. I have been waiting for my best eyesight friend, on the net, to show up, since switching over to Quantum from Chromium. You mentioned in Nov, of last year, that you would start on FF, first thing, in 2018 - I've been checking to see any glimmer of hope. Now, that it is here, I'll be checking back often to see your progress. Any ideas when we can hope to get the absolute apex of accessibility over here at FF. 

Thanks for thinking about us pavel-agarkov!

**Comment by pavel-agarkov** [comment]
@rtlewis1 - I investigated this topic in January as planned, but it turned out to be much harder then I thought. For example FF does would not allow ML to process images that easy. But I still consider porting ML anyway.
Right now I'm working with a back-end for ML and I'm going back to extension right after a first part of it. It is difficult to say exact dates but since you are subscribed to the issue - you will know first.
Thank you for the patience.

**Comment by rtlewis1** [comment]
@pavel-agarkov,

 it is wonderful that you are going to stick with it, for Quantum! One huge difference between ML and the others is that, with your superior extension, the pages don't open in a blinding, sloppy, WHITE mess before settling down to the proper, dark rendering. ML just opens perfectly! I have to literately close my ultra-sensitive eyes as the other, inferior extensions load the page. 

Look forward to hearing from you soon; thank you getting back, my best-eye-friend!

**Comment by rtlewis1** [comment]
@pavel-agarkov 
I know you have been having trouble theming the scrollbar; I have used two templates to do just that. I haven't yet located one of the two; but here is one of them. I wouldn't know if if this could be helpful - so please forgive me if I am being naive here.  Again, thank you for keeping up with with this effort! 

This site will not let me upload a .js file, so here it is:

(custom_scrollbars.uc.js)

```
"use strict";

/* Firefox 57+ userChrome.js tweaks - SCROLLBARS ********************************************** */
/* by Aris (aris-addons@gmx.net)*************************************************************** */
/* Github: https://github.com/aris-t2/customscrollbarsforfx *********************************** */
/* ******************************************************************************************** */

/* ******************************************************************************************** */
/* Custom Scrollbars for Firefox ************************************************************** */
/* version 1.0.4 ****************************************************************************** */
/* ******************************************************************************************** */

/* ***********************************************************************************************

 README
 
 [!] either 'method 1' or 'method 2' are required to prepare Firefox for custom JavaScript files
 [!] 'custom_scrollbars.uc.js' file belongs into Firefox profiles 'chrome' folder!
 -> finding profile folder: address bar > about:profiles > Root Directory > Open Folder
 -> add file to \chrome\ folder (create one, if needed)

 [!] STARTUP CACHE HAS TO BE DELETED AFTER EVERY CHANGE!
 -> finding 'startupCache' folder: address bar > about:profiles > Local Directory > Open Folder > startupCache
 -> close Firefox
 -> delete 'startupCache' folders content

 ENABLING options > set var to true
 DISABLING options > set var to false
 Modifying appearance > change values
 - color - name: red, blue, transparent / hex code: #33CCFF, #FFF
 - color - rgb(a): rgba(0,0,255,0.8) / hsl(a): hsla(240,100%,50%,0.8)
 - numbers: 1, 2, 3 ... 10, 11, 12 ...
 - opacity: 0.0 to 1.0 e.g. 1.4, 1,75
 - gradients: linear-gradient(direction, color, color, color)
 - gradients example: linear-gradient(to right, blue, #33CCFF, rgba(0,0,255,0.8))
 - unsettings predefined gradients: transparent,rgba(255,255,255,0.5),transparent -> transparent,rgba(255,255,255,0.0),transparent
 - button size with arrow (descriptions for vertical scrollbars - analogous for horizontal scrollbars)
 -- 1: space above arrow = arrow height
 -- 1.5: space above arrow = 0.5 * arrow height
 -- 2: no space above arrow
 
 NOTE
 - This is a tiny collection of scrollbar tweaks, not a port of 'NewScrollbars' add-on!
 - Small scrollbar width values will corrupt some parts of the ui!
 
*********************************************************************************************** */

// GENERAL SCROLLBAR SETTINGS
var hide_scrollbars = false; // default: hide_scrollbars = false
var hide_scrollbar_buttons = false; // default: hide_scrollbar_buttons = false
var custom_scrollbar_size = true; // default: custom_scrollbar_size = false
var custom_scrollbar_size_value = 17; // in px // default: custom_scrollbar_size_value = 17
var custom_scrollbar_opacity = false; // default: custom_scrollbar_opacity = false
var custom_opacity_value = "1.0"; // default: custom_opacity_value = "1.0"
// floating scrollbars
var enable_scrollbars_on_top_of_webcontent = false; // default: enable_scrollbars_on_top_of_webcontent = false // uses 'custom_scrollbar_size_value' inside its code
// custom scrollbars
var enable_custom_scrollbars = true; // default: enable_custom_scrollbars = true

// CUSTOM SCROLLBAR SETTINGS ("custom_scrollbar_" --> "cs_")
var cs_thumb_border = 1; // in px // default: cs_thumb_border = 0
var cs_thumb_roundness = 0; // in px // default: cs_thumb_roundness = 0
var cs_buttons_roundness = 0; // in px // default: cs_buttons_roundness = 0
var cs_buttons_as_arrows = false; // default: cs_buttons_as_arrows = false // uses 'custom_scrollbar_size_value' inside its code
var cs_arrows_on_buttons = true; // default: cs_arrows_on_buttons = true // uses 'custom_scrollbar_size_value' inside its code
// button size with arrow - 1: space above arrow = arrow height // 1.5: space above arrow = 0.5 * arrow height // 2: no space above arrow
var cs_arrows_on_buttons_buttons_size = 1.5; // default: cs_arrows_on_buttons_buttons_size = 1.5
// 'flat' scrollbars
var cs_ignore_color_gradients = false; // default: cs_ignore_color_gradients = false

// CUSTOM SCROLLBAR COLORS/GRADIENTS
// - background
var cs_background_color = "#DDDDDD"; // default: cs_background_color = "#DDDDDD"
var cs_background_image_vertical = "linear-gradient(to right,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_background_image_vertical = "unset"
var cs_background_image_horizontal = "linear-gradient(to bottom,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_background_image_horizontal = "unset"
// - corner
var cs_corner_background_color = "#DDDDDD"; // default: cs_corner_background_color = "#DDDDDD"
var cs_corner_background_image = "linear-gradient(45deg,transparent 30%,rgba(255,255,255,0.5) 50%,transparent 70%),linear-gradient(-45deg,transparent 30%,rgba(255,255,255,0.5) 50%,transparent 70%)"; // default: cs_corner_background_image = "unset"
// - thumb/slider
var cs_thumb_color = "#33CCFF"; // default: cs_thumb_color = "#33CCFF"
var cs_thumb_image_vertical = "linear-gradient(to right,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_thumb_image_vertical = "unset"
var cs_thumb_image_horizontal = "linear-gradient(to bottom,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_thumb_image_horizontal = "unset"
var cs_thumb_hover_color = "#66FFFF"; // default: cs_thumb_hover_color = "#66FFFF"
var cs_thumb_hover_image_vertical = "linear-gradient(to right,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_thumb_hover_image_vertical = "unset"
var cs_thumb_hover_image_horizontal = "linear-gradient(to bottom,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_thumb_hover_image_horizontal = "unset"
var cs_thumb_border_color = "#33CCFF"; // default: cs_thumb_border_color = "#33CCFF"
// - buttons
var cs_buttons_color = "#66FFFF"; // default: cs_buttons_color = "#66FFFF"
var cs_buttons_image_vertical = "linear-gradient(to right,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_buttons_image_vertical = "unset"
var cs_buttons_image_horizontal = "linear-gradient(to bottom,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_buttons_image_horizontal = "unset"
var cs_buttons_hover_color = "#33CCFF"; // default: cs_buttons_hover_color = "#33CCFF"
var cs_buttons_hover_image_vertical = "linear-gradient(to right,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_buttons_hover_image_vertical = "unset"
var cs_buttons_hover_image_horizontal = "linear-gradient(to bottom,transparent,rgba(255,255,255,0.5),transparent)"; // default: cs_buttons_hover_image_horizontal = "unset"
// - button arrows
var cs_arrows_on_buttons_color = "#33CCFF"; // default: cs_arrows_on_buttons_color = "#33CCFF"
var cs_arrows_on_buttons_hover_color = "#66FFFF"; // default: cs_arrows_on_buttons_hover_color = "#66FFFF"

// unset background image color gradients -> flat scrollbars
if(cs_ignore_color_gradients==true)
	cs_background_image_vertical
	=cs_background_image_horizontal
	=cs_corner_background_image=cs_thumb_image_vertical
	=cs_thumb_image_horizontal
	=cs_thumb_hover_image_vertical
	=cs_thumb_hover_image_horizontal
	=cs_buttons_image_vertical
	=cs_buttons_image_horizontal
	=cs_buttons_hover_image_vertical
	=cs_buttons_hover_image_horizontal
	="unset";

/* ******************************************************************************************** */
/* ******************************************************************************************** */
/* ******************************************************************************************** */


// Scrollbar code

Components.utils.import("resource://gre/modules/Services.jsm");
var ss =  Components.classes["@mozilla.org/content/style-sheet-service;1"].getService(Components.interfaces.nsIStyleSheetService);

var custom_scrollbars = {

	init: function() {

	var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar, scrollcorner, scrollbar thumb, scrollbar scrollbarbutton {\
	  -moz-appearance: none !important;\
	}\
	scrollbar {\
	  background-color: '+cs_background_color+' !important;\
	}\
	scrollbar[orient="vertical"] {\
	  background-image: '+cs_background_image_vertical+' !important;\
	}\
	scrollbar[orient="horizontal"] {\
	  background-image: '+cs_background_image_horizontal+' !important;\
	}\
	scrollcorner {\
	  background-color: '+cs_corner_background_color+' !important;\
	  background-image: '+cs_corner_background_image+' !important;\
	}\
	scrollbar thumb {\
	  background-color: '+cs_thumb_color+' !important;\
	  border-radius: '+cs_thumb_roundness+'px !important;\
	  border: '+cs_thumb_border+'px solid '+cs_thumb_border_color+' !important;\
	}\
	scrollbar thumb[orient="vertical"] {\
	  background-image: '+cs_thumb_image_vertical+' !important;\
	}\
	scrollbar thumb[orient="horizontal"] {\
	  background-image: '+cs_thumb_image_horizontal+' !important;\
	}\
	scrollbar thumb:hover, scrollbar thumb:active {\
	  background-color: '+cs_thumb_hover_color+' !important;\
	}\
	scrollbar thumb[orient="vertical"]:hover, scrollbar thumb[orient="vertical"]:active {\
	  background-image: '+cs_thumb_hover_image_vertical+' !important;\
	}\
	scrollbar thumb[orient="horizontal"]:hover, scrollbar thumb[orient="horizontal"]:active {\
	  background-image: '+cs_thumb_hover_image_horizontal+' !important;\
	}\
	scrollbar scrollbarbutton {\
	  background-color: '+cs_buttons_color+' !important;\
	  border-radius: '+cs_buttons_roundness+'px !important;\
	}\
	scrollbar[orient="vertical"] scrollbarbutton {\
	  background-image: '+cs_buttons_image_vertical+' !important;\
	}\
	scrollbar[orient="horizontal"] scrollbarbutton {\
	  background-image: '+cs_buttons_image_horizontal+' !important;\
	}\
	scrollbar scrollbarbutton:hover {\
	  background-color: '+cs_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="vertical"] scrollbarbutton:hover {\
	  background-image: '+cs_buttons_hover_image_vertical+' !important;\
	}\
	scrollbar[orient="horizontal"] scrollbarbutton:hover {\
	  background-image: '+cs_buttons_hover_image_horizontal+' !important;\
	}\
	'), null, null);

	ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

	}
};

var cs_scrollbars_scrollbar_button_arrows = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar scrollbarbutton, \
	scrollbar[orient="vertical"] scrollbarbutton,\
	scrollbar[orient="horizontal"] scrollbarbutton, \
	scrollbar[orient="vertical"] scrollbarbutton:hover,\
	scrollbar[orient="horizontal"] scrollbarbutton:hover {\
	  background-color: unset !important;\
	  background-image: unset !important;\
	  border-radius: 0px !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton {\
	  min-height: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  height: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  max-height: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  min-width: '+custom_scrollbar_size_value+'px !important;\
	  width: '+custom_scrollbar_size_value+'px !important;\
	  max-width: '+custom_scrollbar_size_value+'px !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton {\
	  min-width: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  width: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  max-width: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  min-height: '+custom_scrollbar_size_value+'px !important;\
	  height: '+custom_scrollbar_size_value+'px !important;\
	  max-height: '+custom_scrollbar_size_value+'px !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="decrement"] {\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_color+' !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="decrement"]:hover {\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="increment"] {\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_color+' !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="increment"]:hover {\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="decrement"] {\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="decrement"]:hover {\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="increment"] {\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="increment"]:hover {\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid '+cs_buttons_hover_color+' !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

var cs_scrollbars_arrows_on_buttons = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar scrollbarbutton, \
	scrollbar[orient="vertical"] scrollbarbutton,\
	scrollbar[orient="horizontal"] scrollbarbutton, \
	scrollbar[orient="vertical"] scrollbarbutton:hover,\
	scrollbar[orient="horizontal"] scrollbarbutton:hover {\
	  border-radius: 0px !important;\
	}\
	\
	scrollbar[orient="vertical"] > scrollbarbutton {\
	  min-height: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  height: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  max-height: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  min-width: '+custom_scrollbar_size_value+'px !important;\
	  width: '+custom_scrollbar_size_value+'px !important;\
	  max-width: '+custom_scrollbar_size_value+'px !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton {\
	  min-width: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  width: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  max-width: '+(custom_scrollbar_size_value/cs_arrows_on_buttons_buttons_size)+'px !important;\
	  min-height: '+custom_scrollbar_size_value+'px !important;\
	  height: '+custom_scrollbar_size_value+'px !important;\
	  max-height: '+custom_scrollbar_size_value+'px !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="decrement"] {\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_color+' !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="decrement"]:hover {\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="increment"] {\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_color+' !important;\
	}\
	scrollbar[orient="vertical"] > scrollbarbutton[type="increment"]:hover {\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="decrement"] {\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="decrement"]:hover {\
	  border-right: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_hover_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="increment"] {\
	  border-top: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-bottom: '+(custom_scrollbar_size_value/2)+'px solid transparent !important;\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_color+' !important;\
	}\
	scrollbar[orient="horizontal"] > scrollbarbutton[type="increment"]:hover {\
	  border-left: '+(custom_scrollbar_size_value/2)+'px solid '+cs_arrows_on_buttons_hover_color+' !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

var scrollbar_buttons = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar scrollbarbutton {\
	  opacity: 0 !important;\
	}\
	scrollbar[orient="vertical"] scrollbarbutton {\
	  min-height: 1px !important;\
	  height: 1px !important;\
	  max-height: 1px !important;\
	}\
	scrollbar[orient="horizontal"] scrollbarbutton {\
	  min-width: 1px !important;\
	  width: 1px !important;\
	  max-width: 1px !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

var scrollbar_size = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar[orient="vertical"] scrollbarbutton {\
	  min-width: 0 !important;\
	  width: '+custom_scrollbar_size_value+'px !important;\
	  max-width: '+custom_scrollbar_size_value+'px !important;\
	}\
	scrollbar[orient="horizontal"] scrollbarbutton {\
	  min-height: 0 !important;\
	  height: '+custom_scrollbar_size_value+'px !important;\
	  max-height: '+custom_scrollbar_size_value+'px !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

var floating_scrollbars = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar {\
	  position: relative !important;\
	  z-index: 1000000000 !important;\
	}\
	scrollbar[orient="vertical"],\
	scrollbar[orient="horizontal"],\
	scrollbar, scrollcorner {\
	  background-color: transparent !important; \
	  background-image: unset !important; \
	}\
	scrollbar[orient="vertical"] {\
	  -moz-margin-start: -'+custom_scrollbar_size_value+'px !important;\
	  width: '+custom_scrollbar_size_value+'px !important;\
	}\
	scrollbar[orient="horizontal"] {\
	  margin-top: -'+custom_scrollbar_size_value+'px !important;\
	  height: '+custom_scrollbar_size_value+'px !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

var scrollbar_opacity = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar {\
	  opacity: '+custom_opacity_value+' !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

var remove_scrollbars = {

  init: function() {

  var uri = Services.io.newURI("data:text/css;charset=utf-8," + encodeURIComponent('\
	\
	@namespace html url("http://www.w3.org/1999/xhtml");\
	\
	scrollbar, scrollcorner {\
	  display: none !important;\
	  visibility: collapse !important;\
	}\
	\
  '), null, null);

  ss.loadAndRegisterSheet(uri, ss.AGENT_SHEET);

  }
};

// enable settings as configured above
if(enable_custom_scrollbars==true) custom_scrollbars.init();
if(cs_buttons_as_arrows==true && hide_scrollbars==false) cs_scrollbars_scrollbar_button_arrows.init();
if(cs_arrows_on_buttons==true && cs_buttons_as_arrows==false && hide_scrollbars==false) cs_scrollbars_arrows_on_buttons.init();
if(hide_scrollbar_buttons==true) scrollbar_buttons.init();
if(custom_scrollbar_size==true) scrollbar_size.init();
if(enable_scrollbars_on_top_of_webcontent==true) floating_scrollbars.init();
if(custom_scrollbar_opacity==true) scrollbar_opacity.init();
if(hide_scrollbars==true) remove_scrollbars.init();


```

**Comment by pavel-agarkov** [comment]
@rtlewis1 - looks like source code for the method you mentioned located [here](https://github.com/Aris-t2/Scrollbars), but it requires access to local file system which even if it is possible I suppose it will be forbidden by Mozilla. They already rejected my first attempt to publish just for using innerHtml property.

**Comment by pavel-agarkov** [comment]
finally turned out that it is impossible to use this method from WebExtensions.
https://stackoverflow.com/a/36339591

**Comment by rtlewis1** [comment]
Well, that's a bummer, @pavel-agarkov  

It's okay though - we mainly just need out Midnight Lizard to do the wonderful _page_ theming that is comprehensively unmatched anywhere else. Don't give up on the project - it is very much needed at FF!

Thanks again!

**Comment by pavel-agarkov** [comment]
Yes, I understand that it is not the main issue.
I published an experimental version: https://addons.mozilla.org/en-US/firefox/addon/midnight-lizard-quantum
The main problem is that everything works 10 times slower on average.
Do you see the same performance downfall?

**Comment by rtlewis1** [comment]
@pavel-agarkov 

MY MAN - YOU DID IT! 

Yes, there is a little of the slow script anomaly, here and there; but it seems to be improving with cached pages. Maybe that's not it; but I'm pretty sure this represents a close to perfect beginning. Furthermore, I will not be taking this extension off - I've waited way too long for ML on FF. Plus, the main thing is doing what I need - NO BRIGHT-WHITE SLOPPY and BLINDING Page load... YEA!!!

Thank you, tremendously, for this - I will be your dedicated tester, as I am in grad-school.  I will be logging eight to twelve hours per day on ML.

[](http://kb.mozillazine.org/Unresponsive_Script_Warning)

**Comment by pavel-agarkov** [comment]
Thank you, @rtlewis1 
Please be aware that some websites will be completely unusable. For example: YouTube, Facebook, Google Docs and similar heavy-armed websites.
I will try to figure out what exactly causing this.
But I already love the fact that now I can use Midnight Lizard on mobile phone!

**Comment by rtlewis1** [comment]
Awesome! I'll check in tomorrow and give you a user update.

**Comment by rtlewis1** [comment]
@pavel-agarkov 

I have been enjoying my ML on FF for many hours, now, and I am loving it. My academic surfing has been almost as smooth as with ML on Chromium; though the heavy sites still suck a little - but I know it will get better - you are too much of a pro to not figure out how to iron out those FF wrinkles. Everything is now perfectly themed to my Adapta colors  (GTK-Linux). 

Again I'm loving it - thank you. If you need me to do some specific things, just let me know.

![image](https://user-images.githubusercontent.com/37197801/37975572-9f2213b2-31ad-11e8-8765-ffd1ed28ccc8.png)

**Comment by pavel-agarkov** [comment]
I believe I did fix the main part of Firefox specifics and now it works much better. The new version [8.8.4] should arrive shortly. If you find any new bugs please create a new issue or send me email. Thanks a lot.

**Comment by rtlewis1** [comment]
Great to hear @pavel-agarkov , I will do that. 

Also, I'll give you a user update, and get a sufficient review up on the add-ons site by tomorrow in that case. 

Thanks!

**Comment by rtlewis1** [comment]
I just updated to 8.8.4 and it is a fast as lightning, compared to before. You are really talented @pavel-agarkov  - but I've known that since switching to ML, back when it showed up over at Chromium. I'm amazed at your gift, and how terrifically you utilize it (and how fast you competently troubleshoot). 

They say Quantum is the future, and now you are part of it. BTW, FF themes the scrolbars to match the rest the OS's active theme's, app-scrollbars - so I don't see ML-FF having the native theme scrollbars as a problem, whatsoever, it really just seems normal for this ecosystem. As with Chrome/Chromium, this extension will climb to the top, vis-a-vis the number of users. And the folks who switched to Quantum from Chrome/Chromium will be blown away, with delight, because of your contribution.

You made me (and soon many-many others) incredibly happy friend - for the hundredth time THANK YOU!

**Comment by rtlewis1** [comment]
No problems, today, with ML-FF! There really wasn't anything too out of the ordinary - as compared with ML-Chrome/Chromium - everything was smooth, fast, perfectly tweaked - and most importantly - loaded in a soothing manner vis-a-vis my painful Photophobia (hypersensitivity to light). I'll check back in soon, @pavel-agarkov , thanks.

**Comment by rtlewis1** [comment]
Hi, @pavel-agarkov, 

Hope all is well! I was just checking in to say I deeply appreciate all you have done with this extension since we first started communicating; you are genuinely a hero to so many of us. 

Imagine someone who has legs but cannot use them for walking at night, and can barely use them in the daytime. Then you come along and inject some magic into those legs - and now, all-of-a-sudden, they have the ability to walk, run, and even jump.

Thanks a million, as always, brother!
